AX: Live regions should track descendants to handle all aria-relevant values

hoffmanjoshua opened the pull request:
#### 28782b6916c6f0378f75acbf4911925c1c7d2f4f
<pre>
AX: Live regions should track descendants to handle all aria-relevant values
<a href="https://bugs.webkit.org/show_bug.cgi?id=303393">https://bugs.webkit.org/show_bug.cgi?id=303393</a>
<a href="https://rdar.apple.com/165330472">rdar://165330472</a>

Reviewed by Tyler Wilcock.

Previously, we were not tracking descendant changes for aria-atomic=true regions. This meant
that all types of changes were treated as text-changes, rather than also considering node
additions and removals.

To properly support those last two values of aria-relevant, we can keep track of the
descendants of a region, and compute the added/removed/changed announcement more accurately.

Test: accessibility/mac/live-regions/atomic-region-node-added.html

* LayoutTests/accessibility/mac/live-regions/atomic-region-node-added-expected.txt: Added.
* LayoutTests/accessibility/mac/live-regions/atomic-region-node-added.html: Added.
* Source/WebCore/accessibility/AXLiveRegionManager.cpp:
(WebCore::AXLiveRegionManager::handleLiveRegionChange):
(WebCore::AXLiveRegionManager::buildLiveRegionSnapshot const):
(WebCore::AXLiveRegionManager::computeChanges const):
(WebCore::AXLiveRegionManager::computeAnnouncement const):
* Source/WebCore/accessibility/AXLiveRegionManager.h:

Canonical link: <a href="https://commits.webkit.org/303812@main">https://commits.webkit.org/303812@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/f04bf5fbefcebcf796bd557505f26d2ee0cef10a

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows | Apple Internal |
| ----- | ---------------------- | ------- |  ----- |  --------- | ------ |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/133600 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/159/builds/6105 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/138/builds/44792 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/141160 "Built successfully") | [✅ 🛠 win](https://ews-build.webkit.org/#/builders/59/builds/85654 "Built successfully") | [✅ 🛠 ios-apple](https://ews-bridge.webkit.apple.com/builds/sw/T-276/f2c6d2ed-bd8b-4d70-a499-64172ae81bd0/40fad88a-a343-4b14-a783-c92903f40cb3) 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/135470 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/155/builds/6626 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/156/builds/5969 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/102204 "Passed tests") | [✅ 🧪 win-tests](https://ews-build.webkit.org/#/builders/60/builds/69565 "Passed tests") | [✅ 🛠 mac-apple](https://ews-bridge.webkit.apple.com/builds/sw/T-276/f2c6d2ed-bd8b-4d70-a499-64172ae81bd0/d4dac6b7-e8a7-4d13-9fc3-74f1a2860f62) 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/136547 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/162/builds/4695 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/119793 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/83001 "Passed tests") | | [✅ 🛠 vision-apple](https://ews-bridge.webkit.apple.com/builds/sw/T-276/f2c6d2ed-bd8b-4d70-a499-64172ae81bd0/a15a4c2b-1286-465e-a606-871b78814d88) 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/154/builds/4574 "Passed tests") | [✅ 🧪 api-mac-debug](https://ews-build.webkit.org/#/builders/165/builds/2183 "Passed tests") | [⏳ 🛠 wpe-cairo ](https://ews-build.webkit.org/#/builders/WPE-Cairo-Build-EWS "Waiting in queue, processing has not started yet") | | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/113710 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/135/builds/37907 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/143809 "Built successfully") | | 
| | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/153/builds/5777 "Built successfully") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/136/builds/38488 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/110583 "Passed tests") | | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/160/builds/5857 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/161/builds/4954 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/110765 "Passed tests") | | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/28099 "Built successfully and passed tests") | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/164/builds/4432 "Passed tests") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/116046 "Passed tests") | [✅ 🛠 playstation](https://ews-build.webkit.org/#/builders/134/builds/59526 "Built successfully") | | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/158/builds/5829 "Built successfully") | [✅ 🧪 mac-intel-wk2](https://ews-build.webkit.org/#/builders/137/builds/34357 "Passed tests") | | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/157/builds/5677 "Built successfully") | [✅ 🛠 mac-safer-cpp](https://ews-build.webkit.org/#/builders/120/builds/69287 "Built successfully") | | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/163/builds/5920 "Built successfully") | | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/152/builds/5784 "Built successfully") | | | | 
<!--EWS-Status-Bubble-End-->

twilco commented on Source/WebCore/accessibility/AXLiveRegionManager.cpp:
```diff
@@ -186,12 +187,24 @@ LiveRegionSnapshot AXLiveRegionManager::buildLiveRegionSnapshot(AccessibilityObj
+            std::function<void(AccessibilityObject&)> collectDescendants = [&collectDescendants, &descendants] (AccessibilityObject& descendant) {
+                descendants.append(descendant.objectID());
+                for (auto& child : descendant.unignoredChildren())
+                    collectDescendants(downcast<AccessibilityObject>(child.get()));
```
One small optimization — since we know we're going to append all `unignoredChildren()`, could we use `reserveCapacity()` to avoid unnecessary reallocations?

```cpp
const auto& children = descendant.unignoredChildren();
descendants.reserveCapacity(descendants.capacity() + children.size());
for (auto& child : children)
    collectDescendants(downcast<AccessibilityObject>(child.get()));
```

twilco replied:
Is keeping track of every descendant truly necessary? As I read it, the only actual requirement is when any node is added or removed, and that node descends from an `aria-atomic=true`, its whole subtree is announced. Doing one ancestry traversal seems much cheaper than keeping track of entire subtrees. Is there no more efficient way to implement this?

Imagine someone putting aria-atomic on the body of html.spec.whatwg.org — I have to imagine that would totally lock us up with this approach? Obviously that specific case won't happen, but similar ones might and probably already do on other large webpages.

At the very least — do we actually inspect the contents of this vector, or just its size? By my reading, it's just the size, but maybe I missed something. If it's the latter, could we just count the descendants?

hoffmanjoshua replied:
I'm actually in the process of making this algorithm a little more accurate here by actually inspecting AXIDs with a union to see if nodes did change (rather than just compare sizes). I think tracking those changes is important because let's say, for example, someone replaced an entire subtree with a subtree of the same size. We probably don't just want to track counts in that case. 

> Doing one ancestry traversal seems much cheaper than keeping track of entire subtrees. Is there no more efficient way to implement this?

We can definitely do a traversal, but we'd need something to "check" it against to know whether there has been a removal or addition, so we'd still be storing descendants in that case. Did you have another idea on how this could be made efficient?

twilco replied:
> We can definitely do a traversal, but we'd need something to "check" it against to know whether there has been a removal or addition

Sorry, I'm not familiar with the intricacies of the live region rules. My reading of aria-atomic=true is that any change in its subtree, addition or removal, results in the same thing — all descendants are read. So why do we need to do the work of differentiating between these? I'm sure there's just something I don't know about the spec that requires this...just curious what that is 🤔

hoffmanjoshua replied:
Discussed this approach offline with Tyler! For now, we are going to stick with the HashSet approach, but down the line may move to a more fine-grained implementation.

twilco commented on Source/WebCore/accessibility/AXLiveRegionManager.cpp:
```diff
@@ -50,6 +50,7 @@ static constexpr ASCIILiteral accessibilityLanguageAttributeKey = "UIAccessibili
 struct StringLanguagePair {
     String text;
     String language;
+    Vector<AXID> descendants;
```
This struct name no longer seems accurate

twilco commented on LayoutTests/accessibility/mac/live-regions/atomic-region-node-added.html:
```diff
@@ -0,0 +1,44 @@
+
+    var region = document.getElementById("region");
+    var span = document.createElement("span");
+    var text = document.createTextNode(" world");         
```
Are these local variables necessary? Or can they be inlined into the `appendChild` call?

hoffmanjoshua replied:
Inlined them!

twilco commented on LayoutTests/accessibility/mac/live-regions/atomic-region-node-added.html:
```diff
@@ -0,0 +1,44 @@
+if (accessibilityController) {
+    window.jsTestIsAsync = true;
+    accessibilityController.addNotificationListener(notificationCallback);
+    accessibilityController.accessibleElementById("region");
```
Is this necessary? We don't store the result

hoffmanjoshua replied:
Yeah, we need to do this in all of the live-region tests in order to initialize the tree. I've added a comment here in the latest commit explaining that

minorninth commented on Source/WebCore/accessibility/AXLiveRegionManager.cpp:
```diff
@@ -232,9 +245,9 @@ String AXLiveRegionManager::textForObject(AccessibilityObject& object) const
-    // - Additions: New objects
-    // - Deletions: Objects that were removed from the region
-    // - Changes: Text content/values that changed between the same object.
+    // - Additions: New objects, or atomic regions where nodes were added AND text changed.
```
What if a node is added to an atomic live region, but the overall text of the live region doesn't change?

hoffmanjoshua replied:
In that case, we shouldn't speak anything, which aligns with other browsers and platforms

hoffmanjoshua committed:
AX: Live regions should track descendants to handle all aria-relevant values
https://bugs.webkit.org/show_bug.cgi?id=303393
rdar://165330472

Reviewed by Tyler Wilcock.

Previously, we were not tracking descendant changes for aria-atomic=true regions. This meant
that all types of changes were treated as text-changes, rather than also considering node
additions and removals.

To properly support those last two values of aria-relevant, we can keep track of the
descendants of a region, and compute the added/removed/changed announcement more accurately.

Test: accessibility/mac/live-regions/atomic-region-node-added.html

* LayoutTests/accessibility/mac/live-regions/atomic-region-node-added-expected.txt: Added.
* LayoutTests/accessibility/mac/live-regions/atomic-region-node-added.html: Added.
* Source/WebCore/accessibility/AXLiveRegionManager.cpp:
(WebCore::AXLiveRegionManager::handleLiveRegionChange):
(WebCore::AXLiveRegionManager::buildLiveRegionSnapshot const):
(WebCore::AXLiveRegionManager::computeChanges const):
(WebCore::AXLiveRegionManager::computeAnnouncement const):
* Source/WebCore/accessibility/AXLiveRegionManager.h:

Canonical link: https://commits.webkit.org/303812@main
